docs/building-with-stdlib.md

```diff
@@ -0,0 +1,102 @@
+
+We continue fully supporting both models and run CI for both.
+
+## Pros and Cons
+
+### Using trusted/certified Standard Library
+
+Using standard library implementation classes in customer code and on API
+surface instead of `nostd` classes. Certain environments would prefer to have
+**standard** instead of non-standard classes due to security, performance,
+stability and debug'ability considerations. For example, Visual Studio IDE
+provides 1st class Debug experience for Standard containers, plus additional
+runtime-checks for Debug builds that use Standard containers.
+
+### Minimizing binary size
+
+No need to marshal types from standard to `nostd`, then back to standard
+library means less code involved and less memcpy. We use Standard classes
+used elsewhere in the app. We can subsequently improve the process by avoiding
+`KeyValueIterable` transform (and, thus, unnecessary memcpy) when we know
+that the incoming types does not need to be 'transferred' across ABI boundary.
+This is the case for statically linked executables and when SDK is implemented
+as 'header-only' library.
+
+### Avoiding unnecessary extra memcpy (perf improvements)
+
+No need to transform from 'native' standard library types, e.g. `std::map` via
+`KeyValueIterable` means we can bypass that transformation process, if and when
+we know that the ABI compatibility is not a requirement in certain environment.
+ETA perf improvement is 1.5%-3% better perf since an extra transform, memcpy,
+iteration for-each key-value is avoided.
+
+### ABI stability
+
+Obviously this approach does not work in environment where ABI stability
+guarantee must be provided, e.g. for dynamically loadable OpenTelemetry SDK
+module and plugins for products such as NGINX, Envoy, etc.
+
+## Scope of changes needed to implement the feature
+
+### Separate flavors of SDK build
+
+Supported build flavors:
+
+- `nostd` - OpenTelemetry backport of classes for C++11. Not using STD Lib.
+- `stl`   - Standard Library. Best be compiled with C++20 compaitlbe compiler.
+  C++17 may be used with additional dependencies, e.g. MS-GSL or Abseil.
+- `absl`  - TODO: this should allow using Abseil C++ lirary only (no MS-GSL).
+
+### Test setup
+
+Tests to validate that all OpenTelemetry functionality is working the same
+identical way irrespective of what C++ runtime / STL library it is compiled
+with.
+
+### Implementation Details
+
+Feature allows to alias from `nostd::` to `std::` classes for C++17 and above.
+
+Consistent handling of `std::variant` across various OS:
+
+- backport of a few missing variant features, e.g. `std::get` and `std::visit`
+  for older version of Mac OS X. Patches that enable proper
+  handling of `std::visit` and `std::variant` irrespective of OS version
+  to resolve [this quirk](https://stackoverflow.com/questions/52310835/xcode-10-call-to-unavailable-function-stdvisit).
+
+- ability to optionally borrow implementation of C++20 `gsl::span` from
```

Comment:
Makes sense. I'll add this to the write-up.

Appears like we'd only be needing this for:
- Visual Studio 2015 (C++11/14)
- gcc-4.8 (C++14)

Anything gcc-5+ or Visual Studio 2017+ -- won't need this, as it _should_ have `std::span`. I'm doing a runtime feature check though, so if we don't provide `span` implementation, and if the system doesn't have the `<span>` header - build would fail.

Reply:
I'll submit another iteration with detailed table.